components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManagerFeature.kt

```diff
@@ -0,0 +1,61 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.download
+
+import android.Manifest
+import android.annotation.SuppressLint
+import android.support.annotation.RequiresPermission
+import mozilla.components.browser.session.Download
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+
+/**
+ * Feature implementation for proving download functionality for the selected session.
+ */
+class DownloadManagerFeature(
+    private val downloadManager: DownloadManager,
+    private val sessionManager: SessionManager
+) : SessionManager.Observer, Session.Observer {
+
+    private var activeSession: Session? = null
+
+    /**
+     * Starts observing any download on the selected session and send it to the [DownloadManager]
+     * to be processed.
+     */
+    @RequiresPermission(allOf = arrayOf(Manifest.permission.INTERNET, Manifest.permission.WRITE_EXTERNAL_STORAGE))
+    fun start() {
+        activeSession = sessionManager.selectedSession
+        sessionManager.register(this)
+        activeSession?.register(this)
+    }
+
+    /**
+     * Stops observing any download on the selected session.
+     */
+    fun stop() {
+        sessionManager.unregister(this)
+        activeSession?.unregister(this)
+        downloadManager.unregisterListeners()
+    }
+
+    /**
+     * Notifies to the [DownloadManager] that a new download must be processed.
+     */
+    @SuppressLint("MissingPermission")
+    override fun onDownload(session: Session, download: Download): Boolean {
+        downloadManager.download(download)
+        return true
+    }
```

Comment:
We need a way to let the app request the permission only when needed.

I'm thinking of something like this:
* `onDowload()` is getting invoked
* If we do **not** have the permissions:
   * let the app know via a callback that it needs to prompt
   * Return false here so that the `Download` object will be retained
   * Add a method that the app can call when a permission was granted
   * Start the download if we have all permissions now
* If we do have the permissions:
   * Start the download immediately

If this is a too big change then let's work on that in a follow-up issue. We will need that to use such a component in an app like Focus. None of the apps will want to ask the user upfront to accept those permissions. At the time of a download coming in is the best in-context place to ask.

Reply:
What about this?
I add a new callback to the constructor `needToRequestPermissions`,
when I new download comes out and the app doesn't have the permissions I will call it.

When the app gets the permissions granted, the app can call the FeatureDownload.onDownload function with the same parameters that `needToRequestPermissions` gave it. 

What do you think?

```kotlin

class DownloadManagerFeature(
    private val context: Context,
    private val sessionManager: SessionManager,
    private val needToRequestPermissions: (session: Session, download: Download) -> Unit = { _, _ -> },
    private val downloadManager: DownloadManager = DownloadManager(context)
) : SessionManager.Observer, Session.Observer {

/*.....*/

 override fun onDownload(session: Session, download: Download): Boolean {
        return if (context.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE)) {
            downloadManager.download(download)
            true
        } else {
            needToRequestPermissions(session, download)
            false
        }
    }
}
```

Reply:
I wonder if we can make it easier for the app and not require and parameters, something like: `onPermissionsGranted()`. Permission granting is asynchronously and it's quite some overhead for the app to keep a reference to the `Session` and `Download` across potential configuration changes etc.

Could we just check (in the callback) the selected session and see if it has a download attached and then re-trigger this ourselves?

Reply:
Got it! 👍